kazuki-toolkit/src/main/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/Date.kt

```diff
@@ -56,20 +56,76 @@ class DateProperties(private val date: Date) {
 }
 
 class DateFunctions(private val date: Date) {
+    val addYears = function(
+        command = { n: nat ->
+            val nextYear = date.year + n
+            val nextMonth = date.month
+            val nextDay = min(date.day, daysInMonth(nextYear, nextMonth))
+
+            mk_Date(nextYear, nextMonth, nextDay)
+        },
+        post = { n, result -> result == date.functions.addMonths(n * MonthsPerYear) }
+    )
 
-    val addMonths: (integer) -> Date = function(
+    val subtractYears = function(
+        command = { n: nat ->
+            val nextYear = date.year - n
+            val nextMonth = date.month
+            val nextDay = min(date.day, daysInMonth(nextYear, nextMonth))
+
+            mk_Date(nextYear, nextMonth, nextDay)
+        },
+        pre = { n -> date.year > n },
+        post = { n, result -> result == date.functions.subtractMonths(n * MonthsPerYear) }
+    )
+
+    val addMonths: (nat) -> Date = function(
         command = { n ->
-            val nextMonth = (date.month.safeToInt() + n - 1).mod(MonthsPerYear.safeToInt()).toNat() + 1u
-            val yearShift = (date.month.safeToInt() + n - 1).floorDiv(MonthsPerYear.safeToInt())
-            val nextYear = (date.year.safeToInt() + yearShift).toNat()
+            val yearsToAdd = n / MonthsPerYear
+            val monthsToAdd = n % MonthsPerYear
+            val willOverflow = MonthsPerYear < date.month + monthsToAdd
```

Comment:
:pick: under/overflow do make me think of going out of range rather than just the year..
```suggestion
            val willOverflowYear = MonthsPerYear < date.month + monthsToAdd
```

Reply:
True, have renamed.